feat(browser): introduce `toMatchScreenshot` for Visual Regression Testing

macarie opened the pull request:
### Description

This PR introduces **initial support for Visual Regression Testing** for Vitest via a new `toMatchScreenshot` assertion.

**Related issue**: #6265

In this initial iteration:

- The feature supports PNG screenshots and uses `pixelmatch` as the comparator.
- The architecture is extensible: new comparators or codecs can be added easily, as long as they implement the expected interface.
- Comparators and codecs can be asynchronous.
- Screenshot comparison is executed in Node as a browser command, which allows for future adoption of native codecs or comparators.

The logic to get a stable screenshot follows Playwright's approach (with some differences):

1. Uses as baseline an optional reference screenshot or captures a new screenshot.
2. Takes a screenshot and compares it to the baseline.
3. If they match, the page is considered stable and the function returns.
4. If not, it continues with the latest screenshot as the baseline.
5. Repeats until stability is reached or a timeout is hit.

The command has 6 possible outcomes:

| Outcome | Description | Result |
| :------ | :---------- | :----: |
| `#01` | couldn't get a stable screenshot within timeout | `fail` |
| `#02` | stable screenshot taken, but no reference exists and update is not allowed | `fail` |
| `#03` | stable screenshot taken, no reference exists, but updates are allowed | `fail` |
| `#04` | stable screenshot taken on first try and matches reference | `pass` |
| `#05` | stable screenshot taken matches reference after retries | `pass` |
| `#06` | stable screenshot taken doesn't match reference (fallback case) | `fail` |

The client matcher expects an `Element` or `Locator` along with:

- An optional name for the screenshot
- An optional options object that includes:
  - Screenshot options, excluding conflicting ones
  - Comparator name and options
  - Timeout value

#### To-do
- [x] Fix screenshot paths and naming
- [x] Add global config for matcher, including `resolveScreenshotPath` and `resolveDiffPath` functions
- [x] Handle default config in one place
- [x] Use `annotation` API (#7953) to show screenshot paths
- [x] Support thresholds in `pixelmatch` comparator
- [x] Add documentation
- [x] Write test cases

### Please don't delete this checklist! Before submitting the PR, please make sure you do the following:
- [x] It's really useful if your PR references an issue where it is discussed ahead of time. If the feature is substantial or introduces breaking changes without a discussion, PR might be closed.
- [ ] Ideally, include a test that fails without this PR but passes with it.
- [ ] Please, don't make changes to `pnpm-lock.yaml` unless you introduce a new test example.

### Tests
- [ ] Run the tests with `pnpm test:ci`.

### Documentation
- [ ] If you introduce new functionality, document it. You can run documentation with `pnpm run docs` command.

### Changesets
- [ ] Changes in changelog are generated from PR name. Please, make sure that it explains your changes in an understandable manner. Please, prefix changeset messages with `feat:`, `fix:`, `perf:`, `docs:`, or `chore:`.

macarie commented on packages/browser/src/client/tester/expect/toMatchScreenshot.ts:
```diff
@@ -0,0 +1,83 @@
+  const name
+    = typeof nameOrOptions === 'string' ? nameOrOptions : this.currentTestName
+
+  const result = await ensureAwaited(error =>
```
As the matcher itself is async, I thought it would still work if the `expect` is not awaited.

I wanted a safeguard, if possible, as `toMatchSnapshot` is not async, while this one is, so it might be easy to forget to `await` it.

macarie replied:
I ran into an issue here, @sheremet-va.

In `packages/expect/src/jest-extend.ts` (according to the debugger), it checks if `result` is an `instanceof Promise`. But the value returned by `ensureAwaited` is not, so it's assumed to be synchronous. Even if we made it return a proper `Promise`, the code still calls `.then` on it, which breaks the assumption that it was awaited by the user 😕

I'm thinking of dropping this change for now. The fix doesn't seem straightforward, and it's probably outside the scope of this PR. What do you think?

A possible solution might be to return an `ensureAwaited` inside the `.then` that's already being called, this way async matchers would be handled automatically. The downside is that this changes behavior as it would start the promise execution even if it's not explicitly awaited by the user.

macarie commented on packages/browser/src/node/commands/screenshotMatcher/index.ts:
```diff
@@ -0,0 +1,312 @@
+  name: string
+  screenshotOptions: ScreenshotMatcherOptions['screenshotOptions']
+}) {
+  return (screenshot(
```
I agree, but unfortunately a conversion from base64 is needed somewhere, as the comparators usually work with raw bytes.

Do you think we could deprecate the `base64` option in favor of, let's say, `output?: 'buffer' | 'base64'`?

sheremet-va replied:
To return a buffer back to the browser you need to convert it from the buffer to something else, which is again wasteful. And the browser can only receive typed array, the buffer is not implemented in browsers

macarie commented:
I moved the types around to what starts to hopefully make some sense. I created a `shared` folder for the types used in both browser and Node environments.

I've also rebased and resolved the conflicts from main.

macarie commented on packages/browser/jest-dom.d.ts:
```diff
@@ -1,6 +1,7 @@
 // Disable automatic exports.
 
 import { ARIARole } from './aria-role.ts'
+import { ComparatorRegistry, ScreenshotMatcherOptions } from './context.js'
```
I know this creates a circular dependency between this file and `context.d.ts`, was wondering if it's fine tho.

To fix it I would have to move `ScreenshotOptions` out of `context.d.ts` and, as it's using `Locator`, at least half of the other types declared in it 😅

sheremet-va replied:
If it works, it works 😄

macarie commented on packages/vitest/src/node/types/browser.ts:
```diff
@@ -236,6 +300,27 @@ export interface BrowserConfigOptions {
    */
   connectTimeout?: number
+
+  expect?: {
```
These will have to go somewhere else I believe, it was just to make TypeScript happy for now.

macarie commented on packages/browser/src/node/commands/screenshotMatcher/utils.ts:
```diff
@@ -0,0 +1,281 @@
+  const resolvedOptions = deepMerge<GlobalOptions>(
+    Object.create(null),
+    defaultOptions,
+    context.project.config.browser.expect?.toMatchScreenshot ?? {},
```
I don't think these are actually read, but will work on it later, I mostly wanted to move all the defaults and stuff in one place instead of having them spread across the project.

sheremet-va commented on packages/browser/src/node/commands/screenshotMatcher/utils.ts:
```diff
@@ -0,0 +1,281 @@
+/**
+ * Retrieves the name of the browser from a given context.
+ */
+function getBrowserName(context: BrowserCommandContext): string {
```
You can just get the name from `provider.config.browser.name`

sheremet-va commented on packages/browser/context.d.ts:
```diff
@@ -44,6 +45,129 @@ export interface ScreenshotOptions {
   save?: boolean
 }
 
+export interface ComparatorRegistry {
```
Maybe this should have a `Screenshot` as a prefix?

macarie commented on packages/vitest/src/node/config/resolveConfig.ts:
```diff
@@ -56,7 +56,7 @@ function parseInspector(inspect: string | undefined | boolean | number) {
 }
 
-export function resolveApiServerConfig<Options extends ApiConfig & UserConfig>(
+export function resolveApiServerConfig<Options extends ApiConfig & Omit<UserConfig, 'expect'>>(
```
This is provided with the `browser` object, the new `browser.expect` property conflicts with `expect`. I noticed `options.expect` is not used tho, so instead of removing it at runtime I removed it from the type.

macarie commented on packages/vitest/src/node/types/browser.ts:
```diff
@@ -236,6 +236,21 @@ export interface BrowserConfigOptions {
   connectTimeout?: number
+
+  expect?: {
+    toMatchScreenshot?: {
+      [ComparatorName in keyof ToMatchScreenshotComparators]:
+      {
+        /**
+         * The name of the comparator to use for visual diffing.
+         *
+         * @defaultValue `'pixelmatch'`
+         */
+        comparatorName?: ComparatorName
+        comparatorOptions?: ToMatchScreenshotComparators[ComparatorName]
+      }
+    }[keyof ToMatchScreenshotComparators] & ToMatchScreenshotOptions
```
This type is a bit chaotic, but without generics the only solution is to create discriminated unions.

The interface is compatible with `ScreenshotComparatorRegistry`, so I used that to extend it in the providers. In the future this will allow to add more comparators and have them typed correctly in the config by augmenting `ScreenshotComparatorRegistry`.

sheremet-va commented on packages/browser/context.d.ts:
```diff
@@ -1,4 +1,5 @@
 import { SerializedConfig } from 'vitest'
+import type pm from 'pixelmatch'
```
I don't think this is used anywhere in this file

sheremet-va reviewed (commented):
Code looks good to me, but we need a lot more tests for this feature

macarie commented:
I changed the testing approach: instead of relying on pre-generated screenshots, the tests now create them dynamically during execution.

This makes the tests more stable and avoids failures across different OSes, browser versions, or rendering environments. All artifacts are cleaned up automatically after the tests run.

One thing still missing is a test for watch mode. I ran into some issues with it and will take another shot at getting it to work tomorrow.

macarie commented on docs/guide/browser/assertion-api.md:
```diff
@@ -1046,3 +1046,216 @@ await expect.element(queryByTestId('parent')).toHaveSelection('ected text')
+- Font rendering and system fonts
+
+It is recommended to read the
+[Visual Regression Testing guide](/) to
```
🔔 Reminder: change this back once the guide is available

macarie commented on pnpm-lock.yaml:
I'm not sure why there are so many changes here. I tried resetting this file to `main`, but as soon as I add any of the dependencies I use, all these changes show up.

macarie commented on test/browser/specs/to-match-screenshot.test.ts:
Is this the correct location for this test?

macarie commented on docs/guide/browser/assertion-api.md:
````diff
@@ -1046,3 +1046,216 @@ await expect.element(queryByTestId('parent')).toHaveSelection('ected text')
 await expect.element(queryByTestId('next')).toHaveSelection('ne')
 ```
+
+## toMatchScreenshot <Badge type="warning">experimental</Badge> <Version>3.3.0</Version> {#tomatchscreenshot}
````
I guess this version will have to change 😛

Will do before merging, note that this version is written in `config.md` too.

macarie commented:
I will have to debug the failing Webdriverio test on my Windows PC, looks like `arg` is not getting sanitized for some reason. I also don't understand why it fails with `ENOENT` 🙃

macarie commented on packages/browser/src/node/commands/screenshotMatcher/index.ts:
```diff
@@ -0,0 +1,246 @@
+
+  // if there's no reference or if we want to update snapshots, we have to finish the comparison early
+  if (reference === null || updateSnapshot === 'all') {
+    const shouldCreateReference = updateSnapshot !== 'none'
```
Another question from the note: should this check if ref and actual match and eventually set the flag to false? Would it even make sense to update screenshots if they still match?

macarie commented on docs/guide/browser/visual-regression-testing.md:
```diff
@@ -0,0 +1,546 @@
+env:
+  AUTHOR_NAME: 'github-actions[bot]'
+  AUTHOR_EMAIL: '41898282+github-actions[bot]@users.noreply.github.com'
+  COMMIT_MESSAGE: 'test: update visual regression screenshots'
```
todo: add `co-authored-by`

sheremet-va commented on docs/guide/browser/visual-regression-testing.md:
```diff
@@ -0,0 +1,546 @@
+
+For stable tests, use the same environment everywhere. We **strongly recommend**
+cloud services like
+[Microsoft Playwright Testing](https://azure.microsoft.com/en-us/products/playwright-testing)
```
> As new section beneath "Running on GitHub Actions"?

Sounds good to me

macarie replied:
Is this what you had in mind? Maybe I misunderstood, but I think it looks great with the two tabs explaining the two approaches 😄

macarie commented:
Just tested this PR by replacing Playwright's component testing with Vitest in _one of_ our visual regression test suites at work. Pretty impressed with how it went!

The speed difference is huge:

- Vitest: 4m46s
  <img width="1666" height="147" alt="Vitest's result showing it run in 4m46s" src="https://github.com/user-attachments/assets/cedbca63-6dbb-4dce-b972-c7445e1a7012" />

- Playwright: 34m41s
  <img width="1851" height="117" alt="Playwright's result showing it run in 34m41s" src="https://github.com/user-attachments/assets/1ed004c2-a59f-4ad6-8ef5-4577e990e664" />

Not sure why Playwright is taking so long here. This was with just 1 browser instance, we usually run 20 in parallel. Our runs take less than 2 minutes to complete, but the consumed _browser-time_ amounts to ~35m each time, so this result aligns well with what we've always observed. Worth noting: Playwright runs a few more tests, those are non-visual tests and only add 1m40s.

About the failing tests in Vitest:

- half failed because this PR handles some _special_ characters in filenames differently
- the other half looks like an actual bug, I'm gonna dig into this one a bit more. Seems like the tests are running _too fast_ and not waiting for the screenshots to complete. It's happening with our button tests where we capture default and hover states, but the first screenshot is always showing the hover state tho.


Overall tho, really minimal changes needed to get this working!

macarie commented on packages/browser/src/client/tester/expect/toMatchScreenshot.ts:
```diff
@@ -0,0 +1,102 @@
+import type { AsyncExpectationResult, MatcherState } from '@vitest/expect'
```
Now that I checked, this is true, but all the files in `packages/browser/src/client/tester/expect` import from `@vitest/expect`.

macarie commented on test/test-utils/index.ts:
```diff
@@ -209,7 +209,10 @@ async function runCli(command: 'vitest' | 'vite-node', _options?: CliOptions | s
-      await cli.waitForStdout('Waiting for file changes')
+      await Promise.race([
+        cli.waitForStdout('Waiting for file changes'),
+        cli.waitForStdout('Watching for file changes'),
```
I'm testing that the test fails, and the CLI output is this:

```
 DEV  v4.0.0-beta.2 /.../vitest/test/browser/vitest-test-b7af4dc4-e2e5-4bc8-9f78-1ebbbcaa26b2

 × |chromium| basic.test.ts > screenshot-snapshot 165ms
   → expect(element).toMatchScreenshot()

No existing reference screenshot found; a new one was created. Review it before running tests again.

Reference screenshot:
  /.../vitest/test/browser/vitest-test-b7af4dc4-e2e5-4bc8-9f78-1ebbbcaa26b2/__screenshots__/basic.test.ts/screenshot-snapshot-1-chromium-darwin.png

   ❯ ../../../packages/browser/dist/expect-element.js:26:2329 notice
     ↳ Reference screenshot


 Test Files  1 failed (1)
      Tests  1 failed (1)
   Start at  00:15:29
   Duration  761ms (transform 0ms, setup 0ms, collect 9ms, tests 166ms, environment 0ms, prepare 603ms)

 FAIL  Tests failed. Watching for file changes...

```

Am I using the wrong reporter or is the function only meant to run tests that succeed?

sheremet-va replied:
Interesting. It is ok to await stdout then, but the string should include "Tests failed", and let's have a comment like `// waiting for either success or failure`

macarie commented on docs/guide/browser/visual-regression-testing.md:
```diff
@@ -0,0 +1,716 @@
+Microsoft's service shines when you need developers to run visual tests locally.
+
+Maybe you've got designers who want to verify their changes, or developers who
+prefer catching issues before pushing. Yes, it costs money, but compared to the
```
True, I might have gotten a bit too carried away 😄 I love the service and I think it has one of the best price/value ratios ever, solves so many problems.

I toned it down a bit, is it better?

macarie commented on packages/browser/src/node/commands/screenshotMatcher/utils.ts:
```diff
@@ -0,0 +1,266 @@
+        }
+
+        // @ts-expect-error we set the value back right beneath
+        delete this.diffs
```
Done, but I'm re-evaluating the need for a lazy getter here. Supposedly, it's concatenating three strings, do you think it makes sense?

sheremet-va replied:
I don't know 🤷 If they are always used, just return them right away

macarie replied:
They're only used when the test fails, so hopefully most of the time they should not get created. I guess we can keep it, if it becomes an issue it can easily be removed anyway.

macarie commented on packages/browser/src/client/tester/expect/toMatchScreenshot.ts:
```diff
@@ -0,0 +1,102 @@
+    ] satisfies ScreenshotMatcherArguments,
+  )
+
+  if (result.pass === false) {
```
Sorry, not sure I follow here. Is this a _note to self_ or should I change it?

sheremet-va replied:
It is both. I will fix the file path in a separate PR, but I don't know why the images are not visible in UI

sheremet-va replied:
Annotations are now visible in the UI where the `expect().toMatchSnapshot` is called (#8315)

macarie commented on test/browser/fixtures/expect-dom/toMatchScreenshot.test.ts:
```diff
@@ -0,0 +1,303 @@
+  `)
+
+describe('.toMatchScreenshot', () => {
+  test('compares screenshots correctly', async ({ onTestFinished }) => {
```
I added a comment explaining how the tests work above the `describe` and a small explanation of what's going on in the first and second tests.

Let me know if it works.

macarie commented:
Let me know if you don't want the merge commit in the PR's history, I will rebase and remove it once we're done with the review.

Force-pushing after a rebase moves comments up in the "load more" hidden section every time I update from `main` 😄

sheremet-va reviewed (dismissed):
This looks good to me! Thanks for working on this 👍
